### PR TITLE
is_safetensors_compatible refactor

### PR DESCRIPTION
**Pull request by williamberman**
documented, tested, and refactored is_safetensors_compatible

This is my understanding of the logic from reading the function, please double check it!

### CODE REVIEW

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -129,20 +129,52 @@ class AudioPipelineOutput(BaseOutput):
...
-        prefix, raw = os.path.split(pt_filename)
-        if raw == f"pytorch_model{_variant}.bin":
-            # transformers specific
-            sf_filename = os.path.join(prefix, f"model{_variant}.safetensors")
+    """
+    Checking for safetensors compatibility:
+    - By default, all models are saved with the default pytorch serialization, so we use the list of default pytorch
+      files to know which safetensors files are needed.
+    - The model is safetensors compatible only if there is a matching safetensors file for at least one default pytorch
+      file.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pipeline_utils.py:137]: Where do we look for this mapping? If it cannot be explained in one line, no need to include it then.

**Review by sayakpaul — COMMENTED** [review]
Thanks for doing this.

Maybe @Narsil also wants to give this a look?

File: src/diffusers/pipelines/pipeline_utils.py (hunk 1 above)
Comment (Narsil) [reply src/diffusers/pipelines/pipeline_utils.py:137]: > at least one

Isn't that all default pytorch files ?

File: tests/pipelines/test_pipeline_utils.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            "safety_checker/pytorch_model.bin",
+            # Removed: "safety_checker/model.safetensors",
+            "vae/diffusion_pytorch_model.bin",
+            # Removed: "vae/diffusion_pytorch_model.safetensors",
+            "text_encoder/pytorch_model.bin",
+            # Removed: "text_encoder/model.safetensors",
+            "unet/diffusion_pytorch_model.bin",
+            "unet/diffusion_pytorch_model.safetensors",  # <- Only matching safetensors file
+        ]
+        self.assertTrue(is_safetensors_compatible(filenames))
```
Comment (Narsil) [inline tests/pipelines/test_pipeline_utils.py:31]: I was under the impression this shouldn't be true. But I could be wrong.

**Review by Narsil — COMMENTED** [review]
Looks good.

I was under the assumption `diffusers` was only using `safetensors` if **all** files where available in SF format. 

Could be wrong as this seems very deliberate. 


Nice PR !

**Review by pcuenca — COMMENTED** [review]
I was under the same impression as @Narsil (we only downloaded `safetensors` if _all_ files were available in safetensors format). Does this PR change that, or am I just confused?

File: tests/pipelines/test_pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply tests/pipelines/test_pipeline_utils.py:31]: Yes same, I think for now this should not be allowed.

**Comment by patrickvonplaten** [comment]
I think this PR allows combining safetensors and non-safetensors files which is not what we want IMO.

File: tests/pipelines/test_pipeline_utils.py (hunk 2 above)
Comment (williamberman) [reply tests/pipelines/test_pipeline_utils.py:31]: Ah sorry, I mixed up the variant logic with the safetensors logic. We use a variant if there's at least one matching variant file but we only use safetensors if there's a matching safetensor's file for each file. My mistake, will fix!

**Comment by williamberman** [comment]
@pcuenca @patrickvonplaten yep, my mistake will fix! https://github.com/huggingface/diffusers/pull/2499#discussion_r1119111682

File: src/diffusers/pipelines/pipeline_utils.py [hunk 3]
```diff
@@ -129,20 +129,52 @@ class AudioPipelineOutput(BaseOutput):
...
-    for pt_filename in pt_filenames:
-        _variant = f".{variant}" if (variant is not None and variant in pt_filename) else ""
-        prefix, raw = os.path.split(pt_filename)
-        if raw == f"pytorch_model{_variant}.bin":
-            # transformers specific
-            sf_filename = os.path.join(prefix, f"model{_variant}.safetensors")
+    """
+    Checking for safetensors compatibility:
+    - By default, all models are saved with the default pytorch serialization, so we use the list of default pytorch
+      files to know which safetensors files are needed.
```
[the remark this answers is not in the document]
Comment (williamberman) [reply src/diffusers/pipelines/pipeline_utils.py:135]: needed to qualify as being safetensors compatible

File: tests/pipelines/test_pipeline_utils.py (hunk 2 above)
Comment (williamberman) [reply tests/pipelines/test_pipeline_utils.py]: fixed!

**Comment by williamberman** [comment]
> @pcuenca @patrickvonplaten @Narsil yep, my mistake will fix! [#2499 (comment)](https://github.com/huggingface/diffusers/pull/2499#discussion_r1119111682)

Fixed! Sorry for the confusion

**Review by Narsil — APPROVED** [review]
LGTM

**Review by patrickvonplaten — APPROVED** [review]
Nice refactor, LGTM. 

Same comment as Nico though regarding the comma :-)

**Pull request merged**